ipsuite/analysis/model/math.py

```
@@ -22,38 +29,60 @@ def compute_intertia_tensor(centered_positions, masses):
     return I_ab
 
 
-def compute_rot_forces(mol):
+def compute_rot_forces(mol, key: str = "forces"):
     mol_positions = mol.get_positions()
     mol_positions -= mol.get_center_of_mass()
     masses = mol.get_masses()
 
-    f_x_r = np.sum(np.cross(mol.calc.results["forces"], mol_positions), axis=0)
+    # TODO fix for n=2
+    if len(mol) <= 2:
+        result = np.zeros((len(mol), 3))
+        if key == "forces_ensemble":
+            result = result[..., None]
+        return result
+
     I_ab = compute_intertia_tensor(mol_positions, masses)
     I_ab_inv = np.linalg.inv(I_ab)
 
-    mi_ri = masses[:, None] * mol_positions
-    return np.cross(mi_ri, (I_ab_inv @ f_x_r))
+    masses = masses[:, None]
+    mi_ri = masses * mol_positions
+
+    contraction_idxs = "ab, b -> a"
+    if key == "forces_ensemble":
+        mol_positions = mol_positions[..., None]
+        contraction_idxs = "ab, nb -> na"
+
+    f_x_r = np.sum(
+        np.cross(mol.calc.results[key], mol_positions, axisa=1, axisb=1), axis=0
+    )
 
+    # Iinv_fxr = I_ab_inv @ f_x_r but batched for ensembles
+    Iinv_fxr = np.einsum(contraction_idxs, I_ab_inv, f_x_r)
 
-def force_decomposition(atom, mapping):
-    # TODO we should only need to do the mapping once
+    if key == "forces_ensemble":
+        result = np.cross(mi_ri[:, None, :], Iinv_fxr[None, :, :], axisa=2, axisb=2)
+        result = np.transpose(result, (0, 2, 1))
+    else:
+        result = np.cross(mi_ri, Iinv_fxr)
```

might be better to use else if key == forces else raise, if someone passes idk, `ensemble_force` by accident, this would raise an error instead of silently ignore it.